File: quests/system-and-software-design/cache-system-design-base/01-lru/main.c
#include <stdlib.h>
#include <string.h>

#define MAX_KEY 10001

typedef struct Node {
    int key;
    int value;
    struct Node *prev;
    struct Node *next;
} Node;

typedef struct {
    int capacity;
    int count;
    Node *head;
    Node *tail;
    Node **nodeMap;
    Node *nodePool;
    Node *freeNode;
} LRUCache;

LRUCache *lRUCacheCreate(int capacity) {
    LRUCache *cache = malloc(sizeof(LRUCache));
    cache->nodeMap = calloc(MAX_KEY, sizeof(Node *));
    cache->nodePool = malloc(sizeof(Node) * capacity);

    cache->capacity = capacity;
    cache->count = 0;
    cache->head = NULL;
    cache->tail = NULL;
    return cache;
}

void moveToHead(LRUCache *obj, Node *node) {
    // Node is already at start
    if (node == obj->head) {
        return;
    }

    if (node->prev) {
        node->prev->next = node->next;
    }
    if (node->next) {
        node->next->prev = node->prev;
    }

    // Node is last one
    if (node == obj->tail) {
        obj->tail = node->prev;
    }

    node->next = obj->head;
    node->prev = NULL;

    if (obj->head) {
        obj->head->prev = node;
    }
    obj->head = node;

    if (obj->tail == NULL) {
        obj->tail = node;
    }
}

int lRUCacheGet(LRUCache *obj, int key) {
    Node *node = obj->nodeMap[key];

    if (node == NULL) {
        return -1;
    }

    moveToHead(obj, node);
    return node->value;
}

void lRUCachePut(LRUCache *obj, int key, int value) {
    Node *node = obj->nodeMap[key];

    // Key is already there
    if (node != NULL) {
        node->value = value;
        moveToHead(obj, node);
    } else {
        // Key has not been used
        if (obj->count < obj->capacity) {
            // There is enough space
            node = &obj->nodePool[obj->count++];
            node->key = key;
            node->value = value;

            node->next = obj->head;
            node->prev = NULL;

            if (obj->head) {
                obj->head->prev = node;
            }
            obj->head = node;

            if (!obj->tail) {
                obj->tail = node;
            }
            obj->nodeMap[key] = node;
        } else {
            Node *lru = obj->tail;
            obj->nodeMap[lru->key] = NULL;
            lru->key = key;
            lru->value = value;

            obj->nodeMap[key] = lru;
            moveToHead(obj, lru);
        }
    }
}

void lRUCacheFree(LRUCache *obj) {
    if (obj) {
        if (obj->nodeMap)
            free(obj->nodeMap);
        if (obj->nodePool)
            free(obj->nodePool);
        free(obj);
    }
}

/**
 * Your LRUCache struct will be instantiated and called as such:
 * LRUCache* obj = lRUCacheCreate(capacity);
 * int param_1 = lRUCacheGet(obj, key);

 * lRUCachePut(obj, key, value);

 * lRUCacheFree(obj);
*/

File: quests/system-and-software-design/cache-system-design-base/01-lru/main.go
package main

import "fmt"

func main() {
	obj := Constructor(2)
	obj.Put(2, 1)
	obj.PrintNodeList()
	obj.Put(1, 1)
	obj.PrintNodeList()
	obj.Put(2, 3)
	obj.PrintNodeList()
	obj.Put(4, 1)
	obj.PrintNodeList()
	fmt.Println(obj.Get(1))
	fmt.Println(obj.Get(2))
}

func (this *LRUCache) PrintNodeList() {
	if this.startNode != nil {
		node := this.startNode
		fmt.Print(node)
		for node.next != nil {
			node = node.next
			fmt.Print("", node)
		}
	}
	fmt.Println()
}

type LRUCache struct {
	capacity        int
	currentCapacity int
	startNode       *Node
	endNode         *Node
	hashMap         map[int]*Node
}

type Node struct {
	next  *Node
	last  *Node
	key   int
	value int
}

func Constructor(capacity int) LRUCache {
	return LRUCache{
		capacity:        capacity,
		currentCapacity: 0,
		startNode:       nil,
		endNode:         nil,
		hashMap:         make(map[int]*Node),
	}
}

func (this *LRUCache) Get(key int) int {
	node, doesExist := this.hashMap[key]
	if doesExist {
		// node is first node
		if node.last == nil {
			return node.value
		}

		// node is last node
		if node.next == nil {
			node.last.next = nil
			this.endNode = node.last
		} else {
			// node is middle node
			node.last.next = node.next
			node.next.last = node.last
		}

		node.last = nil
		node.next = this.startNode
		this.startNode.last = node
		this.startNode = node
		return node.value
	}
	return -1
}

func (this *LRUCache) Put(key int, value int) {
	node, doesExist := this.hashMap[key]
	if doesExist {
		// node is first node
		if node.last == nil {
			node.value = value
			return
		}

		// node is last node
		if node.next == nil {
			node.last.next = nil
			this.endNode = node.last
		} else {
			// node is middle node
			node.last.next = node.next
			node.next.last = node.last
		}

		node.last = nil
		node.next = this.startNode
		this.startNode.last = node
		this.startNode = node
		this.startNode.value = value
	} else {
		var newNode *Node = &Node{key: key, value: value}

		if this.currentCapacity != 0 {
			this.startNode.last = newNode
		}

		newNode.next = this.startNode
		this.startNode = newNode

		if this.currentCapacity == 0 {
			this.endNode = newNode
		}

		this.hashMap[key] = newNode

		if this.currentCapacity == this.capacity {
			lastNode := this.endNode
			this.endNode.last.next = nil
			this.endNode = this.endNode.last
			delete(this.hashMap, lastNode.key)
		} else {
			this.currentCapacity += 1
		}
	}
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */
